docs/concept/clustering.md

```diff
@@ -0,0 +1,184 @@
+
+### 5.2 Data Sharding
+
+Data distribution across the cluster is determined based on the `shard_num` setting for a group and the specified `entity` in each resource, be it a stream or measure. The resource’s `name` with its `entity` is the sharding key, guiding data distribution to the appropriate Data Node during write operations.
+
+Liaison Nodes retrieve shard mapping information from Meta Nodes to achieve efficient data routing. This information is used to route data to the appropriate Data Nodes based on the sharding key of the data.
+
+This sharding strategy ensures the write load is evenly distributed across the cluster, enhancing write performance and overall system efficiency. BanyanDB uses a hash algorithm for sharding. The hash function maps the sharding key (resource name and entity) to a node in the cluster. Each shard is assigned to the node returned by the hash function.
+
+### 5.3 Data Write Path
+
+Here's a text-based diagram illustrating the data write path in BanyanDB:
+
+```
+User
+ |
+ | API Request (Write)
+ |
+ v
+------------------------------------
+|          Liaison Node             |   <--- Stateless Node, Routes Request
+| (Identifies relevant Data Nodes   |
+|  and dispatches write request)    |
+------------------------------------
+       |                            
+       v                            
+-----------------  -----------------  -----------------
+|  Data Node 1  |  |  Data Node 2  |  |  Data Node 3  |
+|  (Shard 1)    |  |  (Shard 2)    |  |  (Shard 3)    |
+-----------------  -----------------  -----------------
+
+```
+
+1. A user makes an API request to the Liaison Node. This request is a write request, containing the data to be written to the database.
+2. The Liaison Node, which is stateless, identifies the relevant Data Nodes that will store the data based on the entity specified in the request.
+3. The write request is executed across the identified Data Nodes. Each Data Node writes the data to its shard.
+
+This architecture allows BanyanDB to execute write requests efficiently across a distributed system, leveraging the stateless nature and routing/writing capabilities of the Liaison Node, and the distributed storage of Data Nodes.
+
+## 6. Queries in a Cluster
+
+BanyanDB utilizes a distributed architecture that allows for efficient query processing. When a query is made, it is directed to a Query Node.
+
+### 6.1 Query Routing
+
+Query Nodes differ from Liaison Nodes in that they do not store shard mapping information from Meta Nodes. Instead, they access all Data Nodes to retrieve the necessary data for queries. As the query load is lower, it is practical for query nodes to access all data nodes for this purpose. It may increase network traffic, but simplifies scaling out of the cluster.
```

Comment:
It could. Side effects of this solution is there for sure. No matter we wrote or not.

Reply:
This pattern is good at batching and aggregation. Some scenarios can leverage it:

* Ad-hoc "TopN" is a classic aggregation operation. OAP's service and instance TopN metric query will benefit from this pattern.
* "MultiGet" is a batching operation on several entities,  which is natural to retrieve data from all data nodes.
* A Skywalking UI's classic Dashboard will fetch several metrics belonging to an entity. The OAP generates several query operations to fetch data from DB. If it can combine them to issue a batch operation, BanyanDB will perform better than the one with several separate queries.

The scenarios serve as examples to illustrate the potential of "stateless query". As core contributors, it is essential that we possess a thorough understanding of this concept.